File: packages/python/qsu/object/objMerge.py
from ..verify.isObject import isObject


def objMerge(*objects):
	if len(objects) == 0:
		return None

	result = {}

	for source in objects:
		if not isObject(source):
			return None

		for key, value in source.items():
			# Two dicts are merged into a *new* dict, so neither source ends up shared with
			# the result. Everything else, lists included, is replaced whole by the later
			# value. Lodash merges lists index by index instead, which quietly keeps
			# elements the caller meant to drop.
			if isObject(result.get(key)) and isObject(value):
				result[key] = objMerge(result[key], value)
			else:
				result[key] = value

	return result

File: packages/javascript/lib/object/objMerge.ts
import type { AnyValueObject } from '../_types/global.js';
import { isObject } from '../verify/isObject.js';

export function objMerge(...objects: AnyValueObject[]): AnyValueObject | null {
	if (objects.length === 0) {
		return null;
	}

	const result: AnyValueObject = {};

	for (let i = 0, objectsLength = objects.length; i < objectsLength; i += 1) {
		const source = objects[i];

		if (!isObject(source)) {
			return null;
		}

		const keys = Object.keys(source);

		for (let k = 0, keysLength = keys.length; k < keysLength; k += 1) {
			const key = keys[k];
			const value = source[key];

			// Two plain objects are merged into a *new* object, so neither source ends up
			// shared with the result. Everything else, arrays included, is replaced whole by
			// the later value. Lodash merges arrays index by index instead, which quietly
			// keeps elements the caller meant to drop.
			result[key] =
				isObject(result[key]) && isObject(value)
					? (objMerge(result[key], value) as AnyValueObject)
					: value;
		}
	}

	return result;
}
